### ISSUE TITLE
Run-time (re-)binding

### ISSUE THREAD

**Issue by arvind** [uwdata/vega#7]
Support run-time binding/re-binding of:
- [x] data
- [ ] signals

**Comment by floribon** [comment]
Hello. Thanks for the great work on v2, I don't know if I should open a bug but I don't find how to update data after runtime since vega v2.

Here is a test page that I found which works using vega 1 only:
https://gist.github.com/wch/6313003

Basically when I call `view.data(newData)` it crashes in `prototype.data` trying to call `data[k](api)`, while `data` doesn't contain any function. Is late update / dara-refresh a feature that has been temporarily dropped? Thanks

**Comment by arvind** [comment]
Indeed, the API for rebinding data at runtime has changed in v2. [This page](https://github.com/uwdata/vega/wiki/Upgrading-to-2.0) of the documentation should walk you through the process of upgrading your Vega 1 specifications and code to Vega 2.

**Comment by floribon** [comment]
Thanks for the precision, I had missed this one.

Using `remove` then `insert`, I do have my new values but with no transition in between (using some `duration`).

I thought of using `update` instead, but this seem to update only the axis and not the marks.

You can reproduce it in the live editor, on the bar chart example:

```
ved.view.data("table")
  .update(function() { return true; }, "y", function(d) {
    return 10;
  });

ved.view.update();
```

I would expect to see the bar be all the same height of `10`, but they stay the same, only the axis are updated. Is that a known bug? Or am I doing something wrong? Also please let me know if that is not the right place to ask these questions (maybe there's a forum, stack exchange or a way to contact you guys)

PS: FYI In the [Streaming Data](https://github.com/uwdata/vega/wiki/Streaming-Data) doc page, your example uses `.modify` but the function is actually `.update`

**Comment by arvind** [comment]
Thanks for catching the typo, it's been fixed. The height of the bars do not change because the `y` and `y2` properties are defined over the `enter` property set (i.e., only when new data tuples are added). If you were to move those properties to the `update` set instead, the bars are re-rendered as you expect (and appropriately transition if `duration` is specified).

**Comment by floribon** [comment]
Excellent, thanks!

I have two last questions, to be answered when you have time:
1. When trying to update the data of the `stacked_bar` example, only the bar are updated but not the `y`axis. This is likely because it is defined from the `stats` data transform which doesn't seem to be recomputed when `table` changes. Any suggestion to work around that? Can we do the transform at the `scale` definition level? Otherwise a "data" dependency detection mechanism may become useful (or at least re-run all transforms when data is updated)
2. I am trying to update the data of `treemap` during runtime rather than in the specs, but the update/insert methods seem to expect an array of values. Such values are created during the `treemap` transform phase when the data is in the spec, but how to run a transform _before_ sending the data? Or even better, is there a way to feed a new `flare` data that would be transformed since the transform is still in the specs? Maybe it's still a todo for you?

**Comment by arvind** [comment]
Thanks, (1) was indeed a bug. It's been fixed in vega/vega-dataflow@4c10162 and subsequently included in the built files in this repo. Regarding (2), data is actually converted from treejson/topojson/geojson into an array using [datalib](http://vega.github.io/datalib). Essentially, for treejson data, the array [is assembled](https://github.com/vega/datalib/blob/master/src/import/formats/treejson.js) by visiting each element in the tree.
